moveit_ros/moveit_servo/src/servo_calcs.cpp

```diff
@@ -261,80 +245,18 @@ void ServoCalcs::start()
   initial_joint_trajectory->points.push_back(point);
   last_sent_command_ = std::move(initial_joint_trajectory);
 
-  current_state_ = planning_scene_monitor_->getStateMonitor()->getCurrentState();
-  tf_moveit_to_ee_frame_ = current_state_->getGlobalLinkTransform(parameters_->planning_frame).inverse() *
-                           current_state_->getGlobalLinkTransform(parameters_->ee_frame_name);
-  tf_moveit_to_robot_cmd_frame_ = current_state_->getGlobalLinkTransform(parameters_->planning_frame).inverse() *
-                                  current_state_->getGlobalLinkTransform(robot_link_command_frame_);
-
-  stop_requested_ = false;
-  thread_ = std::thread([this] { mainCalcLoop(); });
-  new_input_cmd_ = false;
-}
-
-void ServoCalcs::stop()
-{
-  // Request stop
-  stop_requested_ = true;
-
-  // Notify condition variable in case the thread is blocked on it
+  if (!planning_scene_monitor_->getStateMonitor()->waitForCompleteState(parameters_->move_group_name,
+                                                                        ROBOT_STATE_WAIT_TIME))
   {
-    // scope so the mutex is unlocked after so the thread can continue
-    // and therefore be joinable
-    const std::lock_guard<std::mutex> lock(main_loop_mutex_);
-    new_input_cmd_ = false;
-    input_cv_.notify_all();
+    RCLCPP_ERROR(LOGGER, "Timeout waiting for complete state");
+    throw std::runtime_error("Timeout waiting for complete state");
   }
 
-  // Join the thread
-  if (thread_.joinable())
-  {
-    thread_.join();
-  }
+  updateState();
+  resetLowPassFilters(original_joint_state_);
 }
 
-void ServoCalcs::mainCalcLoop()
-{
-  rclcpp::Rate rate(1.0 / parameters_->publish_period);
```

Comment:
I must have missed this, how does servo keep the rate now?

Reply:
InputResampler.  The diagram I posted should hopefully help clear this up.